### PR TITLE
[ALLUXIO-3316] gRPC

### PR DESCRIPTION
**Pull request by ggezer**

### CODE REVIEW

File: core/common/src/main/java/alluxio/security/authentication/DefaultAuthenticationServer.java [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    return driver;
+  }
+
+  @Override
+  public void registerChannel(UUID channelId, String authorizedUser, SaslServer saslServer) {
+    if (null != mChannels.putIfAbsent(channelId,
+        new AuthenticatedChannelInfo(authorizedUser, saslServer))) {
+      mChannels.remove(channelId);
+      throw new RuntimeException(String
+          .format("Channel: %s already exists in authentication registry.", channelId.toString()));
```
Comment (gpang) [inline core/common/src/main/java/alluxio/security/authentication/DefaultAuthenticationServer.java]: include `authorizedUser` in these messages.

File: core/common/src/main/java/alluxio/security/authentication/DefaultAuthenticationServer.java [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        throw new RuntimeException("Authentication type not supported:" + authType.name());
+    }
+  }
+
+  @Override
+  public List<ServerInterceptor> getInterceptors() {
+    if (!SecurityUtils.isSecurityEnabled()) {
+      return Collections.emptyList();
+    }
+    List<ServerInterceptor> interceptorsList = new ArrayList<>();
```
Comment (gpang) [inline core/common/src/main/java/alluxio/security/authentication/DefaultAuthenticationServer.java]: will this always be size 1?

File: core/common/src/main/java/alluxio/security/authentication/DefaultAuthenticationServer.java [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+      }
+    }
+  }
+
+  /**
+   * Primitive that is invoked periodically for cleaning the registry from clients that has become
+   * stale.
+   */
+  private void cleanupStaleClients() {
+    LocalTime cleanupTime = LocalTime.now();
```
Comment (gpang) [inline core/common/src/main/java/alluxio/security/authentication/DefaultAuthenticationServer.java]: Is local time slower than `System.currentTimeMillis()`?

File: core/common/src/main/java/alluxio/security/authentication/AuthenticationProvider.java [hunk 4]
```diff
@@ -13,6 +13,10 @@
 
 import alluxio.Configuration;
 import alluxio.PropertyKey;
+import alluxio.security.authentication.plain.CustomAuthenticationProvider;
```
[the remark this answers is not in the document]
Comment (ggezer) [reply core/common/src/main/java/alluxio/security/authentication/AuthenticationProvider.java]: They were on the same package.
Each auth scheme has its own subpackage now.

File: core/common/src/main/java/alluxio/security/authentication/DefaultAuthenticationServer.java (hunk 2 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/security/authentication/DefaultAuthenticationServer.java]: Probably never more than 2.

File: core/common/src/main/java/alluxio/security/authentication/DefaultAuthenticationServer.java (hunk 3 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/security/authentication/DefaultAuthenticationServer.java]: I don't understand the question. The reason I used LocalTime was that it has nice tools like addSeconds, isBefore,...

File: core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  private static final int THREADS_MIN = 4;
+
+  public static final ExecutorService ASYNC_CACHE_MANAGER_EXECUTOR =
+      new ThreadPoolExecutor(THREADS_MIN,
+          Configuration.getInt(PropertyKey.WORKER_NETWORK_ASYNC_CACHE_MANAGER_THREADS_MAX),
+          THREAD_STOP_MS, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<>(512),
+          ThreadFactoryUtils.build("AsyncCacheManagerExecutor-%d", true));
+
+  public static final ExecutorService BLOCK_READER_EXECUTOR =
+      new ThreadPoolExecutor(THREADS_MIN + 4,
```
[the remark this answers is not in the document]
Comment (bf8086) [reply core/server/worker/src/main/java/alluxio/worker/grpc/GrpcExecutors.java]: Removed.

File: core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+import java.net.SocketAddress;
+import java.util.concurrent.TimeUnit;
+
+import javax.security.auth.Subject;
+
+/**
+ * Default implementation of {@link BlockWorkerClient}.
+ */
+public class DefaultBlockWorkerClient implements BlockWorkerClient {
+  private static final Logger LOGGER =
```
[the remark this answers is not in the document]
Comment (bf8086) [reply core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java]: Renamed

File: core/client/fs/src/main/java/alluxio/client/block/stream/GrpcDataWriter.java [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+   * @param length the length of the block or file to write, set to Long.MAX_VALUE if unknown
+   * @param chunkSize the chunk size
+   * @param type type of the write request
+   * @param options details of the write request which are constant for all requests
+   * @param client the block worker client
+   */
+  private GrpcDataWriter(FileSystemContext context, final WorkerNetAddress address, long id,
+      long length, long chunkSize, RequestType type, OutStreamOptions options,
+      BlockWorkerClient client) throws IOException {
+    mContext = context;
```
[the remark this answers is not in the document]
Comment (bf8086) [reply core/client/fs/src/main/java/alluxio/client/block/stream/GrpcDataWriter.java]: Thanks, added.

File: core/client/fs/src/main/java/alluxio/client/block/stream/GrpcDataWriter.java [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+        throw new UnavailableException(String.format(
+            "Flush request %s to worker %s is not acked before complete.", writeRequest, mAddress));
+      }
+      posWritten = response.getOffset();
+    } while (mPosToQueue != posWritten);
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mClient.isShutdown()) {
```
[the remark this answers is not in the document]
Comment (bf8086) [reply core/client/fs/src/main/java/alluxio/client/block/stream/GrpcDataWriter.java]: Updated.

File: core/client/fs/src/main/java/alluxio/client/block/stream/LocalFileDataReader.java [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+  /**
+   * Creates an instance of {@link LocalFileDataReader}.
+   *
+   * @param reader the file reader to the block path
+   * @param offset the offset
+   * @param len the length to read
+   * @param chunkSize the chunk size
+   */
+  private LocalFileDataReader(LocalFileBlockReader reader, long offset, long len, long chunkSize)
+      throws IOException {
```
[the remark this answers is not in the document]
Comment (bf8086) [reply core/client/fs/src/main/java/alluxio/client/block/stream/LocalFileDataReader.java]: Removed.

**Pull request merged**